staging/src/k8s.io/apiserver/pkg/server/filters/goaway.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"math/rand"
+	"net/http"
+)
+
+// GoawayDecider decides if server should send a GOAWAY
+type GoawayDecider interface {
+	Goaway(r *http.Request) bool
+}
+
+// WithProbabilisticGoaway returns an http.Handler that send GOAWAY probabilistically
+// according to the given chance for HTTP2 requests. After client receive GOAWAY,
+// the in-flight long-running requests will not be influenced, and the new requests
+// will use a new TCP connection to re-balancing to another server behind the load balance.
+func WithProbabilisticGoaway(inner http.Handler, chance float64) http.Handler {
+	randSource := rand.New(rand.NewSource(1))
+
+	return &goaway{
+		handler: inner,
+		decider: &probabilisticGoawayDecider{
+			chance: chance,
+			next:   randSource.Float64,
```

Comment:
I think this is now unlocked and can't be used from multiple threads. (and if it is locked, it's still a global chokepoint)

I think you need to make a sync.Pool of these; it attempts to solve this problem and is easy. (see: https://golang.org/src/sync/pool.go?s=3925:3957#L114)

Reply:
yes, `rand.Rand` used in multi-goroutines  without locker would panic. I ran a problematic test before with a wrapper `fmt.Println`: `fmt.Println(rnd.Float64())`, which misleading me.

I had already fixed it using a pool to get a rand.Rand, many thanks for your guide.